File: python/geometry.py
from math import *
import numpy as np

#
# Angular geometry
#

#returns theta in range of [-pi,pi)
def standardized_radians(theta):
  return (theta + pi) % (2.*pi) - pi

#returns theta2-theta1 in range of [-pi,pi)
def radians_diff(theta1, theta2):
  return standardized_radians(theta2 - theta1)


#returns theta in range of [-180,180)
def standardized_degrees(theta):
  return (theta + 180) % 360 - 180

#returns theta2-theta1 in range of [-180,180)
def degrees_diff(theta1, theta2):
  return standardized_degrees(theta2 - theta1)


def dot(x1,y1,x2,y2):
  return x1 * x2 + x2 * y2

def length(x,y):
  return sqrt(x**2+y**2)

#
# General geometry
#

def quadratic(a,b,c):
  return (-b +sqrt(b**2. - 4.*a*c))/(2.*a) ,(-b -sqrt(b**2.-4.*a*c))/(2.*a)

def distance(x1,y1,x2,y2):
  return sqrt( (x2-x1)**2 + (y2-y1)**2 )
  
#
# Kinematics
#

# returns first t greater than zero that will reach position x
def time_at_position(x,a,v0,x0=0.):
  x = float(x) - float(x0)
  v0 = float(v0)
  a = float(a)
  if a==0:
    return x/v0
  t1,t2 = quadratic(0.5*a,v0,-x)
  if t1 < 0:
    return t2 if t2 > 0 else None
  if t2 < 0:
    return t1
  return min(t1,t2)
 
  
def velocity_at_time(t, a, v0):
  t = float(t)
  a = float(a)
  v0 = float(v0)
  return v0 + a * t

def velocity_at_position(x, a, v0, x0 = 0):
  a = float(a)
  v0 = float(v0)
  x = float(x) - float(x0)
  t = time_at_position(x=x,v0=v0,a=a)
  v = velocity_at_time(t=t,a=a,v0=v0)
  return v

# below based on http://stackoverflow.com/a/13849249/383967

def unit_vector(v):
  return v / np.linalg.norm(v)

def angle_between(v1, v2):
  """ Returns the angle in radians between vectors 'v1' and 'v2'::

          >>> angle_between((1, 0, 0), (0, 1, 0))
          1.5707963267948966
          >>> angle_between((1, 0, 0), (1, 0, 0))
          0.0
          >>> angle_between((1, 0, 0), (-1, 0, 0))
          3.141592653589793
  """
  v1_u = unit_vector(v1)
  v2_u = unit_vector(v2)
  angle = np.arccos(np.dot(v1_u, v2_u))
  if np.isnan(angle):
      if (v1_u == v2_u).all():
          return 0.0
      else:
          return np.pi
  return angle


# returns an interpolated value from table of k-v
# kv must be sorted by key
def table_lookup(kv, key):
    last = len(kv)-1
    if key <= kv[0][0]:
      return kv[0][1]
    if key >= kv[last][0]:
      return kv[last][1]
    for i in range(0,last):
      if key <= kv[i+1][0]:
        return interpolate(
          key, kv[i][0], kv[i][1], kv[i+1][0], kv[i+1][1])



#returns y for given x based on x1,y1,x2,y2
def interpolate(x, x1, y1, x2, y2):
  x = float(x)
  x1 = float(x1)
  x2 = float(x2)
  y1 = float(y1)
  y2 = float(y2)
  
  #print("x:{} x1:{} y1:{} x2:{} y2:{}".format(x,x1,y1,x2,y2))
  m = (y2 - y1)/( x2 - x1 )
  y = y1 + m * (x-x1)
  return y

File: cpp/geometry.cpp
#include "geometry.h"
#include "iostream"
#include <iomanip>

double interpolate(double x, double x1, double y1, double x2, double y2){
  double m = (y2 - y1)/( x2 - x1 );
  return y1 + m * (x-x1);
}

Point::Point(double x, double y) :
  x(x),y(y){}

string Point::to_string() const {
  stringstream ss;
  ss << std::fixed << std::setprecision(2) << "(" << x << "," << std::setprecision(2) << y << ")";
  return ss.str();
}

Point Point::operator -(Point rhs)
{
  return Point(x-rhs.x,y-rhs.y);
}

string to_string(const Point &p) {
  return p.to_string();
}

Angle Angle::degrees(double d) {
  Angle a;
  a.set_degrees(d);
  return a;
}

Angle Angle::radians(double rad) {
  Angle a;
  a.theta = rad;
  return a;
}

double Angle::radians() const {
  return theta;
}

double Angle::degrees() const  {
  return theta * 180. / M_PI;
}

void Angle::set_degrees(double d) {
  theta = d * M_PI/180.;
}

void Angle::set_radians(double theta_)
{
  theta = theta_;
}

void Angle::standardize() {
  theta = fmod(theta + 99*M_PI , 2.*M_PI) - M_PI;
}

const string Angle::to_string() {
    stringstream s;
    s << degrees() << "°";
    return s.str();
}

bool Angle::operator ==(Angle &rhs)   {

  return theta == rhs.theta;
}

Angle &Angle::operator /=(double d)   {
  theta /= d;
  return *this;
}

Angle Angle::operator /(double d) const  {
  Angle rv;
  rv.theta = this->theta / d;
  return rv;
}

Angle Angle::operator *(double d) const  {
  Angle rv;
  rv.theta = this->theta * d;
  return rv;
}

Angle Angle::operator +(const Angle &rhs) const  {
  Angle rv;
  rv.theta = this->theta + rhs.theta;
  return rv;
}

Angle &Angle::operator +=(const Angle &rhs) {
  theta += rhs.theta;
  return *this;
}

Angle Angle::operator -() {
  return Angle::radians(-theta);
}

Angle Angle::operator -(const Angle &rhs) const  {
  Angle rv;
  rv.theta = this->theta - rhs.theta;
  rv.standardize();
  return rv;
}

double degrees(double radians) {
  return radians * 180 / M_PI;
}

double radians(double degrees) {
  return degrees* M_PI / 180;
}

double standardized_radians(double theta) {
  return fmod(theta + M_PI , 2.*M_PI) - M_PI;
}

double standardized_degrees(double theta) {
  return  fmod(theta + 180., 360) - 180.;
}

double degrees_diff(double theta1, double theta2) {
  return standardized_degrees(theta2 - theta1);
}

double length(double x, double y) {
  return sqrt(x*x+y*y);
}

vector<double> quadratic(double a, double b, double c) {

  return {(-b +sqrt(b*b - 4.*a*c))/(2.*a) ,(-b -sqrt(b*b-4.*a*c))/(2.*a)};
}

double distance(double x1, double y1, double x2, double y2) {
  return length(x2-x1,y2-y1);
}

double distance(Point p1, Point p2) {
  return distance(p1.x,p1.y,p2.x,p2.y);
}

// returns distance of point p from segment from start to end
double distance_from_segment_to_pointt(Point start, Point end, Point p ) {
  double dx = end.x - start.x;
  double dy = end.y - start.y;
  double drx = p.x - start.x;
  double dry = p.y - start.y;

  double progress = (drx * dx + dry * dy)/(dx * dx + dy * dy);
  if(progress < 0)
    return distance(start,p);
  if(progress > 1)
    return distance(end,p);
  double l = length(dx,dy);
  double cte = (dry * dx - drx * dy) / l;
  return fabs(cte);
}


double velocity_at_time(double t, double a, double v0){
  return v0 + a * t;
}

double velocity_at_position(double x, double a, double v0, double x0){
  x = x-x0;
  double t = time_at_position(x,a,v0);
  return velocity_at_time(t,a,v0);
}

Point unit_vector(Point p) {
  auto l=length(p.x,p.y);
  return Point(p.x/l,p.y/l);
}

Angle angle_between(double x1, double y1, double x2, double y2) {
  double dot = x1*x2 + y1*y2; // dot product
  double det = x1*y2 - y1*x2; // determinant
  return Angle::radians( atan2(det, dot) );
}

Angle angle_between(Point p1, Point p2) {
  return angle_between(p1.x,p1.y,p2.x,p2.y);
}

Angle angle_to(Point p1, Point p2) {
  return Angle::radians(atan2(p2.y-p1.y,p2.x-p1.x));
}


double clamp(double value, double min_value, double max_value) {
  if(value < min_value)
    return min_value;
  if (value > max_value)
    return max_value;
  return value;
}

double acceleration_for_distance_and_velocities(double d, double v1, double v2) {
  bool same_signs = (v2 >= 0) == (v1 >= 0);

  double a;
  if(same_signs) {
    a = (v2*v2 - v1*v1) / (2*d);
  } else {
    a = (v2*v2 + v1*v1) / (2*d);
  }

  return a;
}

double time_at_position(double x, double a, double v0, double x0){
  if(isnan(x)) {
    return NAN;
  }
  x = x-x0;

  if(a==0)
    return x/v0;
  auto t = quadratic(0.5*a,v0,-x);
  if (t[0] < 0){
    return t[1] > 0 ? t[1] : NAN;
  }
  if (t[1] < 0){
    return t[0];
  }
  return min(t[0],t[1]);
}

Transform2d Transform2d::pose_to_world_transform(Pose2d pose) {
  Transform2d t = world_to_pose_transform(pose);
  t.t = t.t.inverse();
  t.rotation = -t.rotation;
  
  return t;
}

Transform2d Transform2d::world_to_pose_transform(Pose2d pose) {
  Transform2d t;
  t.t.setIdentity();
  Eigen::Translation2d translation(-pose.position.x, -pose.position.y);
  t.t *= Eigen::Rotation2Dd(-pose.heading);
  t.t *=  translation;
  t.rotation = -pose.heading;

  return t;
}


Point Transform2d::operator() (Point & point) {
  Eigen::Vector2d v;
  v << point.x, point.y;
  Eigen::Vector2d vt = t * v;
  return Point(vt[0], vt[1]);
}

Pose2d Transform2d::operator() (Pose2d & pose) {
  Angle heading = pose.heading + rotation;
  Eigen::Vector2d v;
  v << pose.position.x, pose.position.y;
  Eigen::Vector2d vt = t * v;

  

  
  //double ddx = pose.position.x + dx;
  //double ddy = pose.position.y + dy;
  //double x = ddx * cos(rotation.radians() ) - ddy * sin(rotation.radians());
  //double y = ddx * sin(rotation.radians() ) + ddy * cos(rotation.radians());

  
  return Pose2d(heading, {vt[0],vt[1]});
}

